src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -244,6 +246,231 @@ internal SourceText TryReadFileContent(CommandLineSourceFile file, IList<Diagnos
+
+                allRegexes.Add(config, builder.ToImmutableAndFree());
+            }
+
+            Debug.Assert(allRegexes.Count == analyzerConfigs.Count);
+
+            var optionsBuilder = ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(
+                CaseInsensitiveComparison.Comparer);
+            foreach (var sourceFile in sourcePaths)
+            {
+                var normalizedPath = PathWithForwardSlashSeparators(sourceFile);
+                // The editorconfig paths are sorted from shortest to longest, so matches
+                // are resolved from most nested to least nested, where last setting wins
+                foreach (EditorConfig config in analyzerConfigs)
+                {
+                    if (normalizedPath.StartsWith(config.Directory, StringComparison.Ordinal))
+                    {
+                        string relativePath = normalizedPath.Substring(config.Directory.Length);
+                        ImmutableArray<Regex> regexes = allRegexes[config];
+                        for (int sectionIndex = 0; sectionIndex < regexes.Length; sectionIndex++)
+                        {
+                            if (regexes[sectionIndex].IsMatch(relativePath))
+                            {
+                                var section = config.NamedSections[sectionIndex];
+                                addDiagnosticOptions(section, optionsBuilder);
+                            }
+                        }
+                    }
+                }
+
+                // Avoid extra allocations by using null.
+                allDiagnosticOptions.Add(optionsBuilder.Count > 0 ? optionsBuilder.ToImmutable() : null);
+                optionsBuilder.Clear();
+            }
+
+            allRegexes.Free();
+            Debug.Assert(allDiagnosticOptions.Count == sourcePaths.Count);
+
+            return allDiagnosticOptions.ToImmutableAndFree();
+
+            void addDiagnosticOptions(
+                EditorConfig.Section section,
+                ImmutableDictionary<string, ReportDiagnostic>.Builder builder)
+            {
+                const string DiagnosticOptionPrefix = "dotnet_diagnostic.";
+                const string DiagnosticOptionSuffix = ".severity";
+
+                foreach (var (key, value) in section.Properties)
+                {
+                    // Keys are lowercased in editorconfig parsing
+                    if (key.StartsWith(DiagnosticOptionPrefix, StringComparison.Ordinal) &&
+                        key.EndsWith(DiagnosticOptionSuffix, StringComparison.Ordinal))
+                    {
+                        string diagId = key.Substring(DiagnosticOptionPrefix.Length);
+                        diagId = diagId.Substring(0, diagId.Length - DiagnosticOptionSuffix.Length);
+
+                        ReportDiagnostic? severity = null;
+                        switch (CaseInsensitiveComparison.ToLower(value))
```

Comment:
> switch (CaseInsensitiveComparison.ToLower(value)) [](start = 23, length = 50)

How often are we going to be parsing this? Seems like we're creating an extra allocation per property here. Know it's annoying but may want to consider wiriting out this if by hand.